storage/log_storage.go

```diff
@@ -88,7 +88,13 @@ type LogStorage interface {
 // LeafQueuer provides a write-only interface for the queueing (but not necessarily integration) of leaves.
 type LeafQueuer interface {
 	// QueueLeaves enqueues leaves for later integration into the tree.
-	QueueLeaves(leaves []*trillian.LogLeaf, queueTimestamp time.Time) error
+	// If error is nil, the returned slice of leaves will be the same size as the
+	// input, and each entry will hold:
+	//  - the existing leaf entry if a duplicate has been submitted
+	//  - nil otherwise.
```

Comment:
This API forces the storage implementation to use a single transaction for the batch as it cannot partially succeed for any reason other than duplicate leaves.